delivery_carrier_label_paazl/models/stock_picking.py

```diff
@@ -0,0 +1,227 @@
+
+_logger = logging.getLogger(__name__)
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def action_generate_carrier_label(self):
+        # do nothing for paazl pickings as we have created the labels
+        # during send already
+        return super(
+            StockPicking, self.filtered(lambda x: x.carrier_id.delivery_type != "paazl")
+        ).action_generate_carrier_label()
+
+    def _paazl_send(self):
+        """Register this picking and commit to paazl immediately,
+        return a list of dicts {'exact_price': 'tracking_number':}
+        suitable for delivery.carrier#send_shipping"""
+        result = []
+        api = self._paazl_api()
+        for this in self:
+            if len(this.option_ids) != 1:
+                raise UserError(
+                    _("Picking %s must set exactly one shipping option") % this.name
+                )
+            this._set_a_default_package()
+            status = api.service.order(**this._paazl_order_data())
+            if status.error and status.error.code == 1003:
+                # this order has been sent already, so we update it
+                status = api.service.updateOrder(**this._paazl_order_data())
+            this._paazl_raise_error(status)
+            status = api.service.commitOrder(**this._paazl_commit_order_data())
+            if status.error and status.error.code != 1003:
+                this._paazl_raise_error(status)
+            # we need to create labels in order to get a tracking number
+            status = api.service.generateLabels(**this._paazl_generate_labels_data())
+            this._paazl_raise_error(status)
+            # as we have the labels anyways, attach them
+            self.env["shipping.label"].create(
+                this.get_shipping_label_values(
+                    {
+                        "name": this.name,
+                        "file": base64.b64encode(status.labels),
+                        "file_type": status.labels[:4] == b"%PDF" and "pdf" or "zpl",
+                    }
+                )
+            )
+            result.append(
+                {
+                    # TODO to get this right, we'd have to request shipping options
+                    # and find the price of the shipping option selected
+                    "exact_price": 0,
+                    "tracking_number": status.metaData[0]["trackingNumber"],
+                }
+            )
+        return result
+
+    def _paazl_cancel(self):
+        """Cancel a shipment"""
+        api = self._paazl_api()
+        for this in self:
+            auth_order_id = this._paazl_auth_order_id()
+            status = api.service.cancelShipments(
+                webshop=auth_order_id.pop("webshop"), label=auth_order_id,
+            )
+            this._paazl_raise_error(status)
+
+    def _paazl_api(self):
+        """Return a soap client with the correct url"""
+        # TODO use caching for the wsdl
```

Comment:
Is that as simple as using zeep.CachingClient()? Or are there gotchas in using that?

Reply:
I dislike the idea of having a file system cache, but on the other hand the in memory caching dies with the worker, so it wasn't obvious to me what to do here, hence the todo. https://docs.python-zeep.org/en/master/transport.html#transport-caching looks like it's comparatively simple to write our own caching into some table or similar.

Reply:
In memory caching seems fine to me. But I see this is more elaborate than I thought so you can leave this as a todo.